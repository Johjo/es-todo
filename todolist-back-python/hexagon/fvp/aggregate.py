from abc import ABC, abstractmethod
from collections import OrderedDict
from dataclasses import dataclass

from hexagon.shared.type import TaskKey

@dataclass(frozen=True)
class Task:
    key: TaskKey

    def to_do_the_task(self):
        return DoTheTask(key=self.key)

    def to_choose_the_task(self, other_task: 'Task'):
        return ChooseTheTask(main_task_key=self.key, secondary_task_key=other_task.key)


@dataclass(frozen=True)
class NothingToDo:
    pass


@dataclass(frozen=True)
class DoTheTask:
    key: TaskKey


@dataclass(frozen=True)
class ChooseTheTask:
    main_task_key: TaskKey
    secondary_task_key: TaskKey


@dataclass(frozen=True)
class FvpSnapshot:
    task_priorities: OrderedDict[TaskKey, TaskKey]

    def to_primitive_dict(self) -> dict[TaskKey, TaskKey]:
        return {key: value for key, value in self.task_priorities.items()}

    @classmethod
    def from_primitive_dict(cls, data: dict[TaskKey, TaskKey]) -> 'FvpSnapshot':
        return FvpSnapshot(OrderedDict[TaskKey, TaskKey](data))


class FinalVersionPerfectedSession:
    def __init__(self, task_priorities: OrderedDict[TaskKey, TaskKey]):
        self.task_priorities: OrderedDict[TaskKey, TaskKey] = task_priorities

    def which_task(self, open_tasks: list[Task]) -> NothingToDo | DoTheTask | ChooseTheTask:
        ids = [task.key for task in open_tasks]
        tasks = [task for task in open_tasks if self.task_priorities.get(task.key, None) not in ids]
        match tasks:
            case [task]:
                return task.to_do_the_task()
            case [task_1, task_2, *tasks]:
                return task_1.to_choose_the_task(task_2)
            case _:
                return NothingToDo()

    def choose_and_ignore_task(self, id_chosen: TaskKey, id_ignored: TaskKey) -> None:
        self.task_priorities[id_ignored] = id_chosen

    def reset(self) -> None:
        self.task_priorities = OrderedDict()

    def to_snapshot(self) -> FvpSnapshot:
        return FvpSnapshot(self.task_priorities)

    @classmethod
    def from_snapshot(cls, snapshot: FvpSnapshot) -> 'FinalVersionPerfectedSession':
        return FinalVersionPerfectedSession(task_priorities=snapshot.task_priorities)

    @classmethod
    def create(cls) -> 'FinalVersionPerfectedSession':
        return FinalVersionPerfectedSession(OrderedDict())

    def cancel_priority(self, task_key: TaskKey):
        keys = list()
        for (key, value) in self.task_priorities.items():
            if value == task_key:
                keys.append(key)

        for key in keys:
            del self.task_priorities[key]


class FvpSessionSetPort(ABC):
    @abstractmethod
    def save(self, snapshot: FvpSnapshot) -> None:
        pass

    @abstractmethod
    def by(self) -> FvpSnapshot:
        pass
