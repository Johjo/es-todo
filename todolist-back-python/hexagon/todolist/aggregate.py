from dataclasses import dataclass, replace

from expression import Result, Error, Ok


@dataclass(frozen=True)
class TaskKey:
    value: int


@dataclass
class TaskSnapshot:
    key: TaskKey
    name: str
    is_open: bool


@dataclass
class TodolistSnapshot:
    name: str
    tasks: list[TaskSnapshot]


@dataclass(frozen=True)
class Task:
    key: TaskKey
    name: str
    is_open: bool

    def to_snapshot(self) -> TaskSnapshot:
        return TaskSnapshot(key=self.key, name=self.name, is_open=self.is_open)

    @classmethod
    def from_snapshot(cls, snapshot: TaskSnapshot) -> 'Task':
        return Task(key=snapshot.key, name=snapshot.name, is_open=snapshot.is_open)

    def reword(self, new_name) -> 'Task':
        return replace(self, name=new_name)

    def close_task(self) -> 'Task':
        return replace(self, is_open=False)


@dataclass(frozen=True)
class TodolistAggregate:
    name: str
    tasks: tuple[Task, ...]

    @classmethod
    def create(cls, todolist_name) -> 'TodolistAggregate':
        return TodolistAggregate(name=todolist_name, tasks=())

    @classmethod
    def from_snapshot(cls, snapshot: TodolistSnapshot) -> 'TodolistAggregate':
        return TodolistAggregate(name=snapshot.name, tasks=(*[Task.from_snapshot(task) for task in snapshot.tasks],))

    def to_snapshot(self) -> TodolistSnapshot:
        return TodolistSnapshot(self.name, tasks=[task.to_snapshot() for task in self.tasks])

    def open_task(self, task: Task) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (task,)))

    def close_task(self, key) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")
        return Ok(replace(self, tasks=(*[task.close_task() if task.key == key else task for task in self.tasks],)))

    def reword_task(self, key: TaskKey, new_name: str) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")

        return Ok(replace(self, tasks=(*[task.reword(new_name) if task.key == key else task for task in self.tasks],)))

    def import_tasks(self, task_snapshots: list[TaskSnapshot]) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (*[Task.from_snapshot(snapshot) for snapshot in task_snapshots],)))

