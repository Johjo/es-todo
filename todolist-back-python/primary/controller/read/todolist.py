from abc import ABC, abstractmethod
from dataclasses import dataclass

from dependencies import Dependencies
from hexagon.shared.type import TaskKey, TodolistName, TodolistContext, TodolistContextCount, TaskOpen, TaskName


# todo : use a task key instead of an id
@dataclass
class Task:
    id: TaskKey
    name: TaskName
    is_open: TaskOpen

    def to_dict(self):
        return {
            "id": str(self.id),
            "name": self.name
        }


class TodolistSetReadPort(ABC):
    @abstractmethod
    def task_by(self, todolist_name: str, task_key: TaskKey) -> Task:
        pass

    @abstractmethod
    def all_by_name(self) -> list[TodolistName]:
        pass

    @abstractmethod
    def counts_by_context(self, todolist_name: TodolistName) -> list[tuple[TodolistContext, TodolistContextCount]]:
        pass

    @abstractmethod
    def all_tasks(self, todolist_name: TodolistName) -> list[Task]:
        pass


class TodolistReadController:
    def __init__(self, dependencies: Dependencies):
        self.dependencies = dependencies

    def all_todolist_by_name(self) -> list[str]:
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.all_by_name()

    def task_by(self, todolist_name: str, task_key: TaskKey) -> Task:
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.task_by(todolist_name=todolist_name, task_key=task_key)

    def counts_by_context(self, todolist_name: TodolistName):
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return todolist_set.counts_by_context(todolist_name=todolist_name)

    def export_to_markdown(self):
        return ""

    def to_markdown(self, todolist_name: TodolistName):
        todolist_set: TodolistSetReadPort = self.dependencies.get_adapter(TodolistSetReadPort)
        return to_markdown(todolist_set.all_tasks(todolist_name))


def to_markdown(tasks: list[Task]) -> str:
    def task_to_markdown(task: Task) -> str:
        return f"- [{" " if task.is_open else "x"}] {task.name}"
    return "\n".join([task_to_markdown(task) for task in tasks])
