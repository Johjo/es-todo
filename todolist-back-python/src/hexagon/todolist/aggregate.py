from dataclasses import dataclass, replace

from expression import Result, Error, Ok, Option, Some

from src.hexagon.shared.event import Event, TaskPostponedEvent
from src.hexagon.shared.type import TaskKey, TodolistName, TaskName, TaskExecutionDate, TaskOpen


@dataclass(frozen=True ,eq=True)
class TaskSnapshot:
    key: TaskKey
    name: TaskName
    is_open: TaskOpen
    execution_date: Option[TaskExecutionDate]


@dataclass(frozen=True, eq=True)
class TodolistSnapshot:
    name: TodolistName
    tasks: tuple[TaskSnapshot, ...]


@dataclass(frozen=True, eq=True)
class Task:
    key: TaskKey
    name: TaskName
    is_open: TaskOpen
    execution_date: Option[TaskExecutionDate]

    def to_snapshot(self) -> TaskSnapshot:
        return TaskSnapshot(name=self.name, is_open=self.is_open, key=self.key, execution_date=self.execution_date)

    @classmethod
    def from_snapshot(cls, snapshot: TaskSnapshot) -> 'Task':
        return Task(key=snapshot.key, name=snapshot.name, is_open=snapshot.is_open,
                    execution_date=snapshot.execution_date)

    def reword(self, new_name) -> 'Task':
        return replace(self, name=new_name)

    def close_task(self) -> 'Task':
        return replace(self, is_open=TaskOpen(False))

    def postpone(self, execution_date: TaskExecutionDate):
        return replace(self, execution_date=Some(execution_date))


@dataclass(frozen=True, eq=True)
class TodolistAggregate:
    name: TodolistName
    tasks: tuple[Task, ...]
    events: tuple[Event, ...]

    @classmethod
    def create(cls, todolist_name) -> 'TodolistAggregate':
        return TodolistAggregate(name=todolist_name, tasks=(), events=())

    @classmethod
    def from_snapshot(cls, snapshot: TodolistSnapshot) -> 'TodolistAggregate':
        return TodolistAggregate(name=snapshot.name, tasks=(*[Task.from_snapshot(task) for task in snapshot.tasks],), events=())

    def to_snapshot(self) -> TodolistSnapshot:
        return TodolistSnapshot(self.name, tasks=tuple([task.to_snapshot() for task in self.tasks]))

    def open_task(self, task: Task) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (task,)))

    def close_task(self, key) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")
        return Ok(replace(self, tasks=(*[task.close_task() if task.key == key else task for task in self.tasks],)))

    def reword_task(self, key: TaskKey, new_name: str) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")

        return Ok(replace(self, tasks=tuple([task.reword(new_name) if task.key == key else task for task in self.tasks],)))

    def import_tasks(self, task_snapshots: list[TaskSnapshot]) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (*[Task.from_snapshot(snapshot) for snapshot in task_snapshots],)))

    def postpone_task(self, key: TaskKey, execution_date: TaskExecutionDate) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")

        tasks = tuple([task.postpone(execution_date) if task.key == key else task for task in self.tasks])
        events = self.events + (TaskPostponedEvent(task_key=key, execution_date=execution_date), )
        return Ok(replace(self, tasks=tasks, events=events))

    def uncommited_events(self) -> tuple[Event, ...]:
        return self.events