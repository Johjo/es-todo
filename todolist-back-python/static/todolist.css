/* Global styles */
body {
    margin: 0;
    padding: 0;
    font-family: Arial, sans-serif;
    background-color: #121212;
    color: #ffffff;
}

a {
    color: #1e90ff;
    text-decoration: none;
}

a:hover {
    text-decoration: underline;
}

/* Layout styles */
.menu {
    background-color: #1f1f1f;
    padding: 10px;
    text-align: center;
    border-bottom: 1px solid #333;
}

.menu div a {
    padding: 10px;
    font-size: 16px;
}

.page {
    display: flex;
    flex-direction: row;
    height: calc(100vh - 50px); /* Full height minus menu */
}

.left, .center {
    padding: 20px;
    overflow-y: auto;
}

.left {
    flex: 1;
    max-width: 300px;
    background-color: #1c1c1c;
    border-right: 1px solid #333;
}

.center {
    flex: 3;
    background-color: #181818;
}

/* Task alignment */
.choose-task {
    display: flex;
    gap: 20px;
    margin-top: 20px;
}

.main-task, .secondary-task {
    flex: 1;
    background-color: #1f1f1f;
    border: 1px solid #333;
    border-radius: 5px;
    padding: 15px;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.5);
}

.main-task h2, .secondary-task h2 {
    font-size: 18px;
    margin-bottom: 10px;
    color: #ffcc00;
}

/* Form and buttons */
form {
    margin-top: 15px;
}

input[type="text"], input[type="submit"] {
    padding: 10px;
    margin-top: 10px;
    font-size: 14px;
    border: 1px solid #333;
    border-radius: 5px;
    background-color: #333;
    color: #ffffff;
}

input[type="submit"] {
    cursor: pointer;
}

input[type="submit"]:hover {
    background-color: #444;
}

/* List styles */
ul {
    list-style-type: none;
    padding: 0;
}

li {
    margin-bottom: 10px;
}

li a {
    color: #ff6347;
}

/* Headings */
h2 {
    font-size: 20px;
    margin-bottom: 10px;
    color: #ffcc00;
}

/* Checkbox styles */
input[type="checkbox"] {
    margin-left: 10px;
    accent-color: #1e90ff;
}

/* Scrollbars */
::-webkit-scrollbar {
    width: 8px;
}

::-webkit-scrollbar-thumb {
    background-color: #444;
    border-radius: 4px;
}

::-webkit-scrollbar-thumb:hover {
    background-color: #555;
}

.add-task input[type="text"] {
    width: 90%; /* Le champ occupera toute la largeur du conteneur */

    padding: 10px;
    font-size: 16px;
    border: 1px solid #333;
    border-radius: 5px;
    background-color: #1f1f1f;
    color: #ffffff;
}