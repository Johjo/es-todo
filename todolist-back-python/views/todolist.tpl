<html>
<head>
    <title>Todo</title>
</head>
<body>

{{response}}

Ajouter une t√¢che
<form action="/todo/{{todolist_name}}/item" method="post">
    <input type="text" name="item">
    <input type="submit" value="Add item">
</form>

hello from template
</body>
</html>