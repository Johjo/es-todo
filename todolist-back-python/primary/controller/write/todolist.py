from hexagon.fvp.type import TaskKey
from hexagon.fvp.write.choose_and_ignore_task import ChooseAndIgnoreTaskFvp
from hexagon.todolist.aggregate import TaskSnapshot
from hexagon.todolist.write.close_task import CloseTask
from hexagon.todolist.write.create_todolist import TodolistCreate
from hexagon.todolist.write.import_many_task import ImportManyTask
from hexagon.todolist.write.open_task import OpenTaskUseCase
from hexagon.todolist.write.reword_task import RewordTask
from dependencies import Dependencies
from test.secondary.todolist.test_external_todolist_markdown import MarkdownTodolist


class TodolistWriteController:
    def __init__(self, dependencies: Dependencies) -> None:
        self.dependencies = dependencies

    def create_todolist(self, todolist_name: str) -> None:
        use_case: TodolistCreate = self.dependencies.get_use_case(TodolistCreate)
        use_case.execute(todolist_name=todolist_name)

    def open_task(self, todolist_name: str, task_name: str):
        use_case: OpenTaskUseCase = self.dependencies.get_use_case(OpenTaskUseCase)
        use_case.execute(todolist_name=todolist_name, name=task_name)

    def close_task(self, todolist_name: str, task_key: TaskKey):
        use_case: CloseTask = self.dependencies.get_use_case(CloseTask)
        use_case.execute(todolist_name=todolist_name, key=task_key)

    def reword_task(self, todolist_name: str, task_key: TaskKey, new_name: str):
        use_case: RewordTask = self.dependencies.get_use_case(RewordTask)
        use_case.execute(todolist_name, task_key, new_name)

    def import_many_tasks_from_markdown(self, todolist_name: str, markdown: str):
        use_case: ImportManyTask = self.dependencies.get_use_case(ImportManyTask)
        external_todolist = MarkdownTodolist(markdown)
        use_case.execute(todolist_name, external_todolist)

    def choose_and_ignore_task(self, chosen_task: TaskKey, ignored_task: TaskKey):
        use_case : ChooseAndIgnoreTaskFvp = self.dependencies.get_use_case(ChooseAndIgnoreTaskFvp)
        use_case.execute(chosen_task_id=chosen_task, ignored_task_id=ignored_task)

