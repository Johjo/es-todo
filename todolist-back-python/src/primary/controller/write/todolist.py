from abc import ABC, abstractmethod
from datetime import datetime, timedelta

from src.dependencies import Dependencies
from src.hexagon.fvp.write.reset_fvp_session import ResetFvpSession
from src.hexagon.shared.type import TaskKey, TodolistName, TaskExecutionDate, TaskName
from src.hexagon.fvp.write.cancel_priority import CancelPriority as Fvp_CancelPriority
from src.hexagon.fvp.write.choose_and_ignore_task import ChooseAndIgnoreTaskFvp
from src.hexagon.todolist.write.close_task import CloseTask
from src.hexagon.todolist.write.create_todolist import TodolistCreate
from src.hexagon.todolist.write.import_many_task import ImportManyTask
from src.hexagon.todolist.write.open_task import OpenTaskUseCase
from src.hexagon.todolist.write.postpone_task import PostPoneTask
from src.hexagon.todolist.write.reword_task import RewordTask
from src.secondary.todolist.markdown_todolist import MarkdownTodolist


class DateTimeProviderPort(ABC):
    @abstractmethod
    def now(self) -> datetime:
        pass


class TodolistWriteController:
    def __init__(self, dependencies: Dependencies) -> None:
        self.dependencies = dependencies
        self._datetime_provider = dependencies.get_adapter(DateTimeProviderPort)


    def create_todolist(self, todolist_name: TodolistName) -> None:
        use_case: TodolistCreate = self.dependencies.get_use_case(TodolistCreate)
        use_case.execute(todolist_name=todolist_name)

    def open_task(self, todolist_name: TodolistName, task_name: TaskName):
        use_case: OpenTaskUseCase = self.dependencies.get_use_case(OpenTaskUseCase)
        use_case.execute(todolist_name=todolist_name, name=task_name)

    def close_task(self, todolist_name: TodolistName, task_key: TaskKey):
        use_case: CloseTask = self.dependencies.get_use_case(CloseTask)
        use_case.execute(todolist_name=todolist_name, key=task_key)

    def reword_task(self, todolist_name: TodolistName, task_key: TaskKey, new_name: TaskName):
        use_case: RewordTask = self.dependencies.get_use_case(RewordTask)
        use_case.execute(todolist_name, task_key, new_name)

    def import_many_tasks_from_markdown(self, todolist_name: TodolistName, markdown: str):
        use_case: ImportManyTask = self.dependencies.get_use_case(ImportManyTask)
        external_todolist = MarkdownTodolist(markdown)
        use_case.execute(todolist_name, external_todolist)

    def choose_and_ignore_task(self, chosen_task: TaskKey, ignored_task: TaskKey):
        use_case: ChooseAndIgnoreTaskFvp = self.dependencies.get_use_case(ChooseAndIgnoreTaskFvp)
        use_case.execute(chosen_task_id=chosen_task, ignored_task_id=ignored_task)

    def cancel_priority(self, task_key: TaskKey):
        use_case: Fvp_CancelPriority = self.dependencies.get_use_case(Fvp_CancelPriority)
        use_case.execute(task_key)

    def reset_all_priorities(self) -> None:
        use_case: ResetFvpSession = self.dependencies.get_use_case(ResetFvpSession)
        use_case.execute()

    def postpone_task(self, name: TodolistName, key: TaskKey, execution_date: TaskExecutionDate):
        use_case: PostPoneTask = self.dependencies.get_use_case(PostPoneTask)
        use_case.execute(todolist_name=name, key=key, execution_date=execution_date)

    def postpone_task_to_tomorrow(self, name : TodolistName, key: TaskKey) -> None:
        tomorrow = self._datetime_provider.now() + timedelta(days=1)
        use_case: PostPoneTask = self.dependencies.get_use_case(PostPoneTask)
        use_case.execute(todolist_name=name, key=key, execution_date=TaskExecutionDate(tomorrow.date()))
