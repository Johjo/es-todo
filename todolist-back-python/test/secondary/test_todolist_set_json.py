import json
import os
from dataclasses import replace

import pytest
from approvaltests import verify  # type: ignore
from approvaltests.reporters import PythonNativeReporter  # type: ignore
from expression import Option, Some
from faker import Faker

from hexagon.todolist.aggregate import TodolistSetPort, TodolistSnapshot, TaskSnapshot, TaskKey
from test.hexagon.todolist.fixture import TodolistFaker


class TodolistSetJson(TodolistSetPort):
    def __init__(self, filename: str):
        self._filename = filename

    def save_snapshot(self, todolist: TodolistSnapshot) -> None:
        d = {**(self._load_json()), todolist.name: self._to_todolist_dict(todolist)}

        with open(self._filename, "w") as f:
            f.write(json.dumps(d))

    def _to_todolist_dict(self, todolist: TodolistSnapshot):
        return {"name": todolist.name, "tasks": [self._to_task_dict(task) for task in todolist.tasks]}

    @staticmethod
    def _to_task_dict(task: TaskSnapshot):
        return {"key": task.key.value, "name": task.name, "is_open": task.is_open}

    def _load_json(self):
        try:
            f = open(self._filename, "r")
        except FileNotFoundError:
            d = {}
        else:
            with f:
                d = json.load(f)
        return d

    def by(self, todolist_name: str) -> Option[TodolistSnapshot]:
        d = self._load_json()
        d1 = {**d[todolist_name]}
        return Some(self._todolist_from_dict(d1))

    def _todolist_from_dict(self, d):
        return TodolistSnapshot(name=d["name"], tasks=[self._task_from_dict(task) for task in d["tasks"]])

    @staticmethod
    def _task_from_dict(d):
        return TaskSnapshot(key=TaskKey(d["key"]), name=d["name"], is_open=d["is_open"])


def ensure_directory_exist(path: str ) -> None:
    if not os.path.exists(path):
        os.mkdir(path)


def test_save_todolist():
    fake = TodolistFaker(Faker())
    ensure_directory_exist("data_test")
    filename = "data_test/test_todolist_set_01.json"
    ensure_file_does_not_exist(filename)
    todolist_set = TodolistSetJson(filename)
    todolist_set.save_snapshot(
        replace(fake.a_todolist(), name="my_todolist", tasks=[replace(fake.a_task(1), name="buy milk")]))

    with open(filename, "r") as f:
        verify(f.read(), reporter=PythonNativeReporter())


def ensure_file_does_not_exist(filename):
    if os.path.exists(filename):
        os.remove(filename)


@pytest.fixture
def fake() -> TodolistFaker:
    return TodolistFaker(Faker())


def test_save_two_todolist(fake: TodolistFaker) -> None:
    filename = "data_test/test_todolist_set_02.json"

    ensure_file_does_not_exist(filename)

    todolist_set = TodolistSetJson(filename)
    todolist_set.save_snapshot(replace(fake.a_todolist(), name="todolist_1"))
    todolist_set.save_snapshot(replace(fake.a_todolist(), name="todolist_2"))

    with open(filename, "r") as f:
        verify(f.read(), reporter=PythonNativeReporter())


def test_read_todolist(fake: TodolistFaker) -> None:
    filename = "data_test/test_todolist_set_03.json"

    ensure_file_does_not_exist(filename)

    todolist_set = TodolistSetJson(filename)
    todolist = replace(fake.a_todolist(), name="todolist_1", tasks=[replace(fake.a_task(1), name="buy milk")])

    todolist_set.save_snapshot(todolist)

    assert todolist_set.by(todolist.name) == Some(todolist)
