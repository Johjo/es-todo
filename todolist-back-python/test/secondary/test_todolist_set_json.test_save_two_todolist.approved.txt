{"todolist_1": {"name": "todolist_1", "tasks": []}, "todolist_2": {"name": "todolist_2", "tasks": []}}
